grpc.go

```diff
@@ -1069,14 +1079,18 @@ func (a *agentGRPC) CloseStdin(ctx context.Context, req *pb.CloseStdinRequest) (
 		return emptyResp, err
 	}
 
-	var file *os.File
-	if proc.termMaster != nil {
-		file = proc.termMaster
-	} else {
-		file = proc.stdin
+	// If stdin is nil, which can be the case when using a terminal,
+	// there is nothing to do.
+	if proc.stdin == nil {
+		return emptyResp, nil
 	}
 
-	if err := file.Close(); err != nil {
+	proc.Lock()
+	defer proc.Unlock()
+
+	proc.stdinClosed = true
+
+	if err := proc.stdin.Close(); err != nil {
```

Comment:
Only set stdinClosed to true when proc.stdin.Close() succeeds? Otherwise we might end up failing `CloseStdin()` and also start to ignore `WriteStdin()`. It's protected by the mutex so it's safe to just move `proc.stdinClosed = true` to after `stdin.Close()`.

Reply:
I have thought about it and I was not sure, but that's fair, let's set the flag only is this succeeded.